internal/abi/fastly/kvstore_guest.go

```diff
@@ -10,166 +10,343 @@ import (
-	if body.h == invalidBodyHandle {
-		return nil, FastlyError{Status: FastlyStatusNone}
+	result := KVLookupResult{
+		Body:       &body,
+		Meta:       meta.AsBytes(),
+		Generation: uint32(generation),
 	}
 
-	return &body, nil
+	return result, nil
 }
 
 // witx:
+//          (@interface func (export "insert")
+//              (param $store $kv_store_handle)
+//              (param $key string)
+//              (param $body_handle $body_handle)
+//              (param $insert_config_mask $kv_insert_config_options)
+//              (param $insert_configuration (@witx pointer $kv_insert_config))
+//              (param $handle_out (@witx pointer $kv_store_insert_handle))
+//              (result $err (expected (error $fastly_status)))
+//          )
 //
-//  (@interface func (export "insert")
-//	  (param $store $object_store_handle)
-//	  (param $key string)
-//	  (param $body_handle $body_handle)
-//	  (result $err (expected (error $fastly_status)))
-//	)
-
-//go:wasmimport fastly_object_store insert
+
+//go:wasmimport fastly_kv_store insert
 //go:noescape
-func fastlyObjectStoreInsert(
-	h objectStoreHandle,
+func fastlyKVStoreInsert(
+	h kvstoreHandle,
 	keyData prim.Pointer[prim.U8], keyLen prim.Usize,
 	b bodyHandle,
+	mask kvInsertConfigMask,
+	config prim.Pointer[kvInsertConfig],
+	insertHandle prim.Pointer[kvstoreInsertHandle],
 ) FastlyStatus
 
-// Insert adds a key/value pair to the kv store.
-func (o *KVStore) Insert(key string, value io.Reader) error {
+// Insert returns a handle to a pending key/value pair insertion.
+func (k *KVStore) Insert(key string, value io.Reader) (kvstoreInsertHandle, error) {
 	body, err := NewHTTPBody()
 	if err != nil {
-		return err
+		return 0, err
 	}
 
 	if _, err := io.Copy(body, value); err != nil {
-		return err
+		return 0, err
 	}
 
 	keyBuffer := prim.NewReadBufferFromString(key).Wstring()
 
-	if err := fastlyObjectStoreInsert(
-		o.h,
+	var mask kvInsertConfigMask
+	var config kvInsertConfig
+
+	var insertHandle kvstoreInsertHandle = invalidKVInsertHandle
+
+	if err := fastlyKVStoreInsert(
+		k.h,
 		keyBuffer.Data, keyBuffer.Len,
 		body.h,
+		mask,
+		prim.ToPointer(&config),
+		prim.ToPointer(&insertHandle),
+	).toError(); err != nil {
+		return 0, err
+	}
+
+	if insertHandle == invalidKVInsertHandle {
+		return 0, FastlyStatusError.toError()
+	}
+
+	return insertHandle, nil
+}
+
+// witx:
+//
+//          (@interface func (export "insert_wait")
+//              (param $handle $kv_store_insert_handle)
+//              (param $kv_error_out (@witx pointer $kv_error))
+//              (result $err (expected (error $fastly_status)))
+//          )
+
+//go:wasmimport fastly_kv_store insert_wait
+//go:noescape
+func fastlyKVStoreInsertWait(
+	h kvstoreInsertHandle,
+	kvErr prim.Pointer[KVError],
+) FastlyStatus
+
+// InsertWait returns the status of the given pending insertion handle.
+func (kv *KVStore) InsertWait(inserth kvstoreInsertHandle) error {
```

Comment:
Total 🚲 🏠:

```suggestion
func (kv *KVStore) InsertWait(insertH kvstoreInsertHandle) error {
```
Capitalizing the H looks less weird to me.

Reply:
Then I should probably go and make the the rest of them `H`-andle too.